### PR TITLE
Bluetooth: rpmsg: bt_rpmsg_close() implementation 

### LINKED ISSUE #52907: Bluetooth: Add functionality for bt_disable for off-chip HCI drivers
**Is your feature request related to a problem? Please describe.**
At the moment bt_disable cannot be used with rpmsg, spi, uart etc. 

**Describe the solution you'd like**
Currently bt_hci_driver.open just sets up communication for off-chip HCI drivers, this would also have to also set up the controller on this command. A bt_hci_driver.close functionality would have to be implemented that tears down the controller.

### PR DESCRIPTION
**Pull request by pawelosyp**
Added bt_rpmsg_close() implementation, any comments appreciated, addressing #52907 
In short, function invokes reset CMD on network core, cleans stuff related to IPC and puts network core in reset.
Checking for freed endpoints changed due to situation I observed: when network core creates two endpoints, the IPC on the app core has listed endpoints with names even if they are not bounded.

Fixes #52907

### CODE REVIEW

**Comment by hermabe** [comment]
Some thoughts:
- Some of this is very nrf53 specific. Is it possible to make it more generic? Not sure if anyone else is using this driver, but anyway.
- Sending HCI Reset is a good start. Some applications may want to do a full teardown of the Bluetooth subsystem. Maybe we should add a function `bt_disable_raw` and call that on the netcore when `bt_disable` is called on the app core? This would need some changes to the hci_rpmsg sample. It does also not have to be done in this PR.

I am not very familiar with the ipc stuff, mostly speaking from Bluetooth perspective.

**Comment by jori-nordic** [comment]
I agree with @hermabe , could you put all the nrf53-specific functions in an `nrf53_support.c` file? it can be in the same dir, and you can use `target_sources_ifdef(` with the right nrf53 define to compile it.

@hermabe I don't get your second point, the PR will put the whole netcore in reset, so no need to teardown the bluetooth hci_raw driver (on the netcore). `bt_disable` already calls the `.close` method of the HCI driver.

File: drivers/bluetooth/hci/rpmsg.c [hunk 1]
```diff
@@ -317,9 +325,55 @@ static int bt_rpmsg_open(void)
...
+	if (!buf) {
+		LOG_ERR("Allocating command buffer failed!");
+		return -ENOBUFS;
+	}
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_RESET, NULL, &rsp);
+	if (err) {
+		LOG_ERR("Sending reset command failed with: %d", err);
+		return err;
+	}
```
Comment (Thalley) [inline drivers/bluetooth/hci/rpmsg.c:344]: Does the other drivers also send an HCI reset on close?

**Review by jori-nordic — DISMISSED** [review]

**Comment by jori-nordic** [comment]
@carlocaione pinging to know if the vrings changes are fine

File: subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c [hunk 2]
```diff
@@ -95,7 +95,7 @@ static bool check_endpoints_freed(struct ipc_rpmsg_instance *rpmsg_inst)
 	for (size_t i = 0; i < NUM_ENDPOINTS; i++) {
 		rpmsg_ept = &rpmsg_inst->endpoint[i];
 
-		if (strcmp("", rpmsg_ept->name) != 0) {
+		if (rpmsg_ept->bound == true) {
```
Comment (carlocaione) [inline subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:98]: Tagging @jacksonc-xlnx

IUC this is slightly changing the semantics though.

- `rpmsg_ept->bound == true` means that the EPT is bounded on the remote and local endpoint
- `strcmp("", rpmsg_ept->name) != 0` means that the EPT was advertised by the local endpoint (but not necessarily bounded yet)

Comment (pawelosyp) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:98]: This change was made due to firmware on network core I am working with. In my case net FW creates two endpoints and the main app connects to one or another endpoint (in each situation it uses only one endpoint). In such case I was not able to successfully call `ipc_service_close_instance(...)` when the `check_endpoints_freed(...)` was checking by names.

Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:98]: Did you call the `.deregister_endpoint` function on both the endpoints before closing the instance down?

Comment (pawelosyp) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:98]: Only on the one that is bonded. For the second one I would need to register it first to have any handle for deregister. I have not found another way to execute 'ipc_service_close_instance' (one it to register and deregister unused endpoint on application side and the second one is to check for freed not by name but by bounded flag). 
If that change is problematic we can solve it on our side by registering and deregistering ept.

Comment (carlocaione) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:98]: The other option is to add a `FORCE` parameter to `ipc_service_close_instance` to basically forcefully closing all the endpoints associated to that instance.

Comment (pawelosyp) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:98]: I am not very familiar with IPC but in my opinion endpoints that are not bound should not block closing API because they are not used by API on that side. From my point of view checking freed by bound flag was most elegant. Especially if the endpoint was registered by the second side.

Comment (carlescufi) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:98]: @jacksonc-xlnx ping

Comment (jacksonc-xlnx) [reply subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:98]: This all sounds reasonable - and that looks like a good change. I would say that it belongs in a separate commit to the other changes however, as it is seemingly unrelated.

**Review by jori-nordic — DISMISSED** [review]

**Comment by jori-nordic** [comment]
Did a rebase and ran downstream tests for bt enable / disable, it seems to work ok.
@pawelosyp are you fine with the changes? I rewrote the commit message.

**Comment by pawelosyp** [comment]
@jori-nordic For me, everything looks OK. On our side, we have put bodies of cpunet_forceon/off() functions inside the rpmsg driver. But it works the same way. Thank you for reopening.

**Review by carlocaione — DISMISSED** [review]

**Review by Thalley — DISMISSED** [review]

File: drivers/bluetooth/hci/rpmsg.c [hunk 3]
```diff
@@ -317,9 +325,55 @@ static int bt_rpmsg_open(void)
 	return 0;
 }
 
+static int bt_rpmsg_close(void)
+{
+	int err;
+	struct net_buf *buf = NULL;
+	struct net_buf *rsp = NULL;
```
Comment (jhedberg) [inline drivers/bluetooth/hci/rpmsg.c:332]: Unnecessary initialisation upon declaration (just remove the `= NULL`

File: drivers/bluetooth/hci/rpmsg.c [hunk 4]
```diff
@@ -317,9 +325,55 @@ static int bt_rpmsg_open(void)
...
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_RESET, 0);
+	if (!buf) {
+		LOG_ERR("Allocating command buffer failed!");
+		return -ENOBUFS;
+	}
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_RESET, NULL, &rsp);
```
Comment (jhedberg) [inline drivers/bluetooth/hci/rpmsg.c:340]: There's a buffer leak here, since you're actually not doing anything with the allocated `buf` (not even calling `net_buf_unref()` on it). You actually don't need the `buf` and `rsp` buffers at all if you just intend to send a command without additional parameter and don't care about the contents of the command complete event either. You can simply do:
```c
err = bt_hci_cmd_send_sync(BT_HCI_OP_RESET, NULL, NULL);
```

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/hci_core.c [hunk 5]
```diff
@@ -4081,6 +4081,9 @@ int bt_disable(void)
 	 */
 	atomic_clear_bit(bt_dev.flags, BT_DEV_ENABLE);
 
+	/* Set to initial value */
+	hci_cmd_pool.uninit_count = CONFIG_BT_BUF_CMD_TX_COUNT;
```
Comment (jhedberg) [inline subsys/bluetooth/host/hci_core.c:4085]: Move this change to its own commit (still in this same PR), since it's not specific to the rpmsg driver.

Comment (jhedberg) [reply subsys/bluetooth/host/hci_core.c:4085]: Actually, why is this change needed? net_buf pools should be self-managing as long as you're correctly allocating and freeing (unreffing) buffers.

Comment (jhedberg) [reply subsys/bluetooth/host/hci_core.c:4085]: Or to put it another way `net_buf_poo.uninit_count` should be seen as an internal variable to the net_buf_pool implementation - you should never need to touch it from outside of the net_buf subsystem.

File: drivers/bluetooth/hci/rpmsg.c (hunk 1 above)
Comment (jhedberg) [reply drivers/bluetooth/hci/rpmsg.c:344]: > Is there a reason why we, for just this driver, are sending an HCI reset when closing the driver? Should we not do it here? Should we do it for the other drivers?
> 
> Just seems odd to have this change in behavior from driver to driver

I agree. We could potentially also control this using driver-specific quirk flags, so that the host-side is responsible for doing the reset if the driver claims that it needs it.

File: subsys/bluetooth/host/hci_core.c (hunk 5 above)
Comment (pawelosyp) [reply subsys/bluetooth/host/hci_core.c:4085]: @jhedberg I'll address all your comments today later. It's hard to say why I did that (I was working on it in March) as I am checking it now it looks like this structure is initialized statically. As I remember there was a wrong initial value after reenabling BT.

File: drivers/bluetooth/hci/rpmsg.c (hunk 1 above)
Comment (pawelosyp) [reply drivers/bluetooth/hci/rpmsg.c]: @Thalley @jhedberg This implementation was based on the ST one. I don't know what ST's radio is doing on custom reset. For me sending an HCI reset before power cycling makes this operation less brutforce.

File: drivers/bluetooth/hci/rpmsg.c (hunk 3 above)
Comment (pawelosyp) [reply drivers/bluetooth/hci/rpmsg.c]: Removed.

File: drivers/bluetooth/hci/rpmsg.c (hunk 4 above)
Comment (pawelosyp) [reply drivers/bluetooth/hci/rpmsg.c]: Good point, reworked it as you said.

File: subsys/bluetooth/host/hci_core.c (hunk 5 above)
Comment (pawelosyp) [reply subsys/bluetooth/host/hci_core.c]: @jhedberg One more time checked that and now it seems to work without touching hci_cmd_pool. This modification is removed.

**Review by jhedberg — DISMISSED** [review]

File: drivers/bluetooth/hci/rpmsg.c [hunk 6]
```diff
@@ -317,9 +325,42 @@ static int bt_rpmsg_open(void)
...
+		DEVICE_DT_GET(DT_CHOSEN(zephyr_bt_hci_rpmsg_ipc));
+
+	err = ipc_service_close_instance(hci_ipc_instance);
+	if (err) {
+		LOG_ERR("Closing IPC service failed with: %d", err);
+		return err;
+	}
+
+#ifdef CONFIG_SOC_NRF5340_CPUAPP
+	cpunet_forceoff();
```
Comment (jori-nordic) [inline drivers/bluetooth/hci/rpmsg.c:354]: Thinking about it, this might be a problem if other stuff's running on the network core, like say zigbee or a custom protocol. They'll just get an abrupt communication link break.

What I propose is that this gets merged, and then I'll make another PR, specific to 53, that enables/disables the forcing on and off of the netcore.

That way advanced users that have something else on the netcore can just disable that.

**Review by jori-nordic — DISMISSED** [review]

File: drivers/bluetooth/hci/rpmsg.c (hunk 6 above)
Comment (pawelosyp) [reply drivers/bluetooth/hci/rpmsg.c:354]: I think that we can go that way.

In the situation you are describing I see another problem - the `bt_rpmsg_close(void)` do not know about another IPC endpoints so if there is any opened the `ipc_service_close_instance(hci_ipc_instance)` will fail.

File: drivers/bluetooth/hci/rpmsg.c [hunk 7]
```diff
@@ -294,6 +298,10 @@ static int bt_rpmsg_open(void)
 	const struct device *hci_ipc_instance =
 		DEVICE_DT_GET(DT_CHOSEN(zephyr_bt_hci_rpmsg_ipc));
 
+#ifdef CONFIG_SOC_NRF5340_CPUAPP
+	cpunet_forceon();
```
Comment (carlescufi) [inline drivers/bluetooth/hci/rpmsg.c:302]: Please add weak functions instead of the Kconfig `#ifdef`:
See here:
https://github.com/zephyrproject-rtos/zephyr/blob/29bbcb1e38f84b505c7fe7e3f7ab4bb3497eea4c/drivers/bluetooth/hci/h4.c#L498

Also, the name needs to change. Please use the same one as above, `bt_hci_transport_setup`.

File: drivers/bluetooth/hci/rpmsg.c (hunk 6 above)
Comment (carlescufi) [inline drivers/bluetooth/hci/rpmsg.c:354]: Also use a weak function as I suggested for the setup and name the weak function, and name it `bt_hci_transport_teardown`.

**Review by carlescufi — CHANGES_REQUESTED** [review]

File: drivers/bluetooth/hci/rpmsg.c [hunk 8]
```diff
@@ -288,12 +288,28 @@ static struct ipc_ept_cfg hci_ept_cfg = {
 	},
 };
 
+int __weak bt_hci_transport_setup(const struct device *dev)
+{
+	return 0;
```
Comment (jori-nordic) [inline drivers/bluetooth/hci/rpmsg.c:293]: might need an `ARG_UNUSED(dev)` here to appease the compiler warnings

**Review by jori-nordic — DISMISSED** [review]

**Review by jhedberg — DISMISSED** [review]

**Review by jori-nordic — APPROVED** [review]

File: drivers/bluetooth/hci/rpmsg.c (hunk 7 above)
Comment (pawelosyp) [reply drivers/bluetooth/hci/rpmsg.c]: @carlescufi Both comments addressed.

**Review by carlescufi — APPROVED** [review]
Thanks for your patience!
